File: frontend/src/lib/portfolio.ts
import { customFetch } from "@/api/mutator";

export type PortfolioReviewAction = {
  code: string;
  severity?: string;
  headline: string;
  summary?: string | null;
  recommendation?: string | null;
};

export type PortfolioRationale = {
  position_key?: string | null;
  strategy?: string | null;
  why?: string | null;
  entry_plan?: string | null;
  profit_take_plan?: string | null;
  risk_plan?: string | null;
  roll_or_reopen_plan?: string | null;
  tags: string[];
  updated_at?: string | null;
};

export type PortfolioPosition = {
  position_key: string;
  as_of?: string | null;
  source_row_ref?: string | null;
  ticker: string;
  instrument_type?: string | null;
  strategy?: string | null;
  option_side?: string | null;
  quantity?: number | null;
  expiration?: string | null;
  strike?: number | null;
  cost_basis?: number | null;
  mark?: number | null;
  unrealized_pnl?: number | null;
  unrealized_pnl_pct?: number | null;
  dte?: number | null;
  status?: string | null;
  latest_flags: PortfolioReviewAction[];
  needs_rationale?: boolean;
  rationale_updated_at?: string | null;
};

export type PortfolioPositionDetail = PortfolioPosition & {
  rationale?: PortfolioRationale | null;
  rationale_history: PortfolioRationale[];
  latest_review_id?: string | null;
  latest_review_summary_markdown?: string | null;
};

export type PortfolioReview = {
  id: string;
  date?: string | null;
  generated_at?: string | null;
  summary_markdown?: string | null;
  actions: PortfolioReviewAction[];
  position_keys: string[];
};

export type PortfolioSyncResult = {
  ok: boolean;
  enqueued: boolean;
  job_id: string;
  run_id?: string | null;
};

export type PortfolioRationaleUpdate = {
  strategy?: string | null;
  why?: string | null;
  entry_plan?: string | null;
  profit_take_plan?: string | null;
  risk_plan?: string | null;
  roll_or_reopen_plan?: string | null;
  tags: string[];
};

export async function fetchPortfolioPositions(): Promise<PortfolioPosition[]> {
  const response = await customFetch<{ data: PortfolioPosition[] }>(
    "/api/v1/portfolio/positions",
    { method: "GET" },
  );
  return sortPortfolioPositions(response.data);
}

export async function fetchPortfolioPositionDetail(
  positionKey: string,
): Promise<PortfolioPositionDetail> {
  const response = await customFetch<{ data: PortfolioPositionDetail }>(
    `/api/v1/portfolio/positions/${encodeURIComponent(positionKey)}`,
    { method: "GET" },
  );
  return response.data;
}

export async function fetchPortfolioReviews(): Promise<PortfolioReview[]> {
  const response = await customFetch<{ data: PortfolioReview[] }>(
    "/api/v1/portfolio/reviews",
    { method: "GET" },
  );
  return response.data;
}

export async function updatePortfolioRationale(
  positionKey: string,
  payload: PortfolioRationaleUpdate,
): Promise<PortfolioPositionDetail> {
  const response = await customFetch<{ data: PortfolioPositionDetail }>(
    `/api/v1/portfolio/positions/${encodeURIComponent(positionKey)}/rationale`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    },
  );
  return response.data;
}

export async function syncPortfolioNow(): Promise<PortfolioSyncResult> {
  const response = await customFetch<{ data: PortfolioSyncResult }>("/api/v1/portfolio/sync", {
    method: "POST",
  });
  return response.data;
}

export function matchesPortfolioPositionSearch(
  position: PortfolioPosition,
  searchTerm: string,
): boolean {
  const normalized = searchTerm.trim().toLowerCase();
  if (!normalized) return true;
  return [
    position.ticker,
    position.position_key,
    position.strategy ?? "",
    position.option_side ?? "",
    position.status ?? "",
    ...(position.latest_flags ?? []).map((flag) => `${flag.code} ${flag.headline}`),
  ].some((value) => value.toLowerCase().includes(normalized));
}

export function sortPortfolioPositions(
  positions: PortfolioPosition[],
): PortfolioPosition[] {
  return [...positions].sort((left, right) => {
    const leftScore = Number(Boolean(left.needs_rationale)) * 10 + Number(Boolean(left.latest_flags?.length));
    const rightScore =
      Number(Boolean(right.needs_rationale)) * 10 + Number(Boolean(right.latest_flags?.length));
    if (leftScore !== rightScore) {
      return rightScore - leftScore;
    }
    const tickerCompare = left.ticker.localeCompare(right.ticker);
    if (tickerCompare !== 0) return tickerCompare;
    return left.position_key.localeCompare(right.position_key);
  });
}

export function parsePortfolioTags(value: string): string[] {
  return value
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean);
}

export function formatPortfolioTags(tags: string[] | null | undefined): string {
  return (tags ?? []).join(", ");
}

File: backend/app/schemas/portfolio.py
"""Schemas for shared-workspace portfolio browsing and rationale editing."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlmodel import Field, SQLModel


class PortfolioReviewActionRead(SQLModel):
    """One review action or flag attached to a position or daily review."""

    code: str
    severity: str = "info"
    headline: str
    summary: str | None = None
    recommendation: str | None = None


class PortfolioRationaleRead(SQLModel):
    """Current durable rationale for one position."""

    position_key: str | None = None
    rolled_from_position_key: str | None = None
    strategy: str | None = None
    why: str | None = None
    entry_plan: str | None = None
    profit_take_plan: str | None = None
    risk_plan: str | None = None
    roll_or_reopen_plan: str | None = None
    tags: list[str] = Field(default_factory=list)
    updated_at: datetime | None = None


class PortfolioRationaleHistoryRead(PortfolioRationaleRead):
    """Historical rationale version."""


class PortfolioRationaleUpdate(SQLModel):
    """Payload for creating or updating a position rationale."""

    strategy: str | None = None
    why: str | None = None
    entry_plan: str | None = None
    profit_take_plan: str | None = None
    risk_plan: str | None = None
    roll_or_reopen_plan: str | None = None
    tags: list[str] = Field(default_factory=list)


class PortfolioPositionRead(SQLModel):
    """List-friendly normalized position view."""

    position_key: str
    as_of: datetime | None = None
    source_row_ref: str | None = None
    ticker: str
    instrument_type: str | None = None
    strategy: str | None = None
    option_side: str | None = None
    quantity: float | None = None
    expiration: str | None = None
    strike: float | None = None
    cost_basis: float | None = None
    mark: float | None = None
    unrealized_pnl: float | None = None
    unrealized_pnl_pct: float | None = None
    dte: int | None = None
    status: str | None = None
    latest_flags: list[PortfolioReviewActionRead] = Field(default_factory=list)
    needs_rationale: bool = False
    rationale_updated_at: datetime | None = None


class PortfolioPositionDetailRead(PortfolioPositionRead):
    """Detailed position payload for the portfolio detail pane."""

    rationale: PortfolioRationaleRead | None = None
    rationale_history: list[PortfolioRationaleHistoryRead] = Field(default_factory=list)
    latest_review_id: str | None = None
    latest_review_summary_markdown: str | None = None


class PortfolioReviewRead(SQLModel):
    """Daily portfolio review payload."""

    id: str
    date: str | None = None
    generated_at: datetime | None = None
    summary_markdown: str | None = None
    actions: list[PortfolioReviewActionRead] = Field(default_factory=list)
    position_keys: list[str] = Field(default_factory=list)


class PortfolioSyncRead(SQLModel):
    """Result of enqueuing a manual portfolio sync."""

    ok: bool = True
    enqueued: bool = True
    job_id: str
    run_id: str | None = None


class PortfolioReviewRunRequest(SQLModel):
    """Raw Google Sheet rows for server-side portfolio review generation."""

    positions_rows: list[list[Any]] = Field(default_factory=list)
    trades_rows: list[list[Any]] = Field(default_factory=list)
    generated_at: str | None = None


class PortfolioReviewRunResult(SQLModel):
    """Result of running the portfolio review engine (cron / manual)."""

    ok: bool = True
    generated_at: str
    position_count: int = 0
    flagged_count: int = 0
    missing_rationale_count: int = 0
    rolls_detected_count: int = 0
    review_id: str
    snapshot_path: str
    review_json_path: str
    review_markdown_path: str


class PortfolioRollEventRead(SQLModel):
    """One detected or recorded option roll."""

    id: str
    rolled_from_position_key: str
    rolled_to_position_key: str
    rolled_at: datetime
    net_credit_cents: int = 0
    source_trade_ids: list[str] = Field(default_factory=list)
    status: str = "auto_carried"
    confidence: float | None = None
